File: black_scholes.cpp
/*
 * Black-Scholes European Call Option Pricer & Greeks
 * ====================================================
 * Implémente la CDF et PDF de la loi normale standard
 * sans dépendance externe (uniquement <cmath>).
 */
#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

#ifndef M_SQRT1_2
#define M_SQRT1_2 0.70710678118654752440
#endif

#include <cmath>
#include <iostream>
#include <iomanip>
#include <string>
using namespace std;
// ====================================================================
//  Distribution normale standard
// ====================================================================

/**
 * PDF de la loi normale standard : φ(x) = (1/√2π) e^{-x²/2}
 */
double norm_pdf(double x) {
    return (1.0 / sqrt(2.0 * M_PI)) * exp(-0.5 * x * x);
}

/**
 * CDF de la loi normale standard via l'approximation d'Abramowitz & Stegun.
 * Précision : |erreur| < 7.5 × 10⁻⁸
 *
 * On utilise erfc (complementary error function) de <cmath> :
 *   Φ(x) = 0.5 × erfc(-x / √2)
 */
double norm_cdf(double x) {
    return 0.5 * erfc(-x * M_SQRT1_2);
}

// ====================================================================
//  Classe BlackScholes
// ====================================================================

class BlackScholes {
public:
    double S, K, T, r, sigma;
    double d_plus, d_minus;

    /**
     * Constructeur
     * @param S     Prix du sous-jacent
     * @param K     Prix d'exercice (strike)
     * @param T     Temps à l'échéance (années)
     * @param r     Taux sans risque (annualisé)
     * @param sigma Volatilité (annualisée)
     */
    BlackScholes(double S, double K, double T, double r, double sigma)
        : S(S), K(K), T(T), r(r), sigma(sigma)
    {
        double sqrtT = sqrt(T);
        d_plus  = (log(S / K) + (r + 0.5 * sigma * sigma) * T)
                  / (sigma * sqrtT);
        d_minus = d_plus - sigma * sqrtT;
    }

    // === Prix ===

    /** C = S Φ(d₊) - K e^{-rT} Φ(d₋) */
    double price() const {
        return S * norm_cdf(d_plus)
             - K * exp(-r * T) * norm_cdf(d_minus);
    }

    // === Greeks ===

    /** Δ = Φ(d₊) */
    double delta() const {
        return norm_cdf(d_plus);
    }

    /** Γ = φ(d₊) / [σ S √T] */
    double gamma() const {
        return norm_pdf(d_plus) / (sigma * S * sqrt(T));
    }

    /** V = S √T φ(d₊) */
    double vega() const {
        return S * sqrt(T) * norm_pdf(d_plus);
    }

    /** Θ = -σS φ(d₊) / [2√T] - rK e^{-rT} Φ(d₋)   (par an) */
    double theta() const {
        double term1 = -sigma * S * norm_pdf(d_plus) / (2.0 * std::sqrt(T));
        double term2 = -r * K * exp(-r * T) * norm_cdf(d_minus);
        return term1 + term2;
    }

    /** ρ = K T e^{-rT} Φ(d₋) */
    double rho() const {
        return K * T * exp(-r * T) * norm_cdf(d_minus);
    }

    // === Affichage ===

    void print_summary() const {
        auto row = [](const std::string& label, double value) {
            std::cout << "  " << left << setw(20) << label
                      << "= " << right <<setw(12)
                      << fixed << setprecision(6) << value
                      << endl;
        };

        cout << string(55, '=') <<endl;
        cout << "  BLACK-SCHOLES : European Call Option Pricer (C++)" <<endl;
        cout << string(55, '=') <<endl;
        cout << "  S = " << S << "  |  K = " << K << "  |  T = " << T <<endl;
        cout << "  r = " << r << "  |  sigma = " << sigma <<endl;
        cout << string(55, '-') << endl;
        row("d+", d_plus);
        row("d-", d_minus);
        cout << std::string(55, '-') <<endl;
        row("Prix du Call", price());
        row("Delta", delta());
        row("Gamma", gamma());
        row("Vega", vega());
        row("Theta (par an)", theta());
        row("Theta (par jour)", theta() / 365.0);
        row("Rho", rho());
        cout << string(55, '=') <<endl;
    }
};

// ====================================================================
//  Main : exemple d'utilisation
// ====================================================================

int main() {
    
    // test 
   

    // Paramètres
    double S     = 100.0;   // Spot
    double K     = 100.0;   // Strike
    double T     = 1.0;     // Maturité (1 an)
    double r     = 0.05;    // Taux sans risque (5%)
    double sigma = 0.2;     // Volatilité (20%)

    BlackScholes bs(S, K, T, r, sigma);
    
    bs.print_summary();
    

    return 0;
}

File: black_scholes.py
"""
Black-Scholes European Call Option Pricer & Greeks
====================================================

Formules dérivées dans les notes de cours :

    C   = S Φ(d₊) - K e^{-r(T-t)} Φ(d₋)

    d₊  = [ln(S/K) + (r + σ²/2)(T-t)] / [σ√(T-t)]
    d₋  = d₊ - σ√(T-t)

Greeks :
    Δ (Delta) = Φ(d₊)
    Γ (Gamma) = φ(d₊) / [σ S √(T-t)]
    V (Vega)  = S √(T-t) φ(d₊)
    Θ (Theta) = -σS φ(d₊) / [2√(T-t)] - rK e^{-r(T-t)} Φ(d₋)
    ρ (Rho)   = K(T-t) e^{-r(T-t)} Φ(d₋)

où Φ = CDF normale standard, φ = PDF normale standard.
"""

import math
from scipy.stats import norm


class BlackScholes:
    """Pricer Black-Scholes pour un call européen avec tous les grecs."""

    def __init__(self, S: float, K: float, T: float, r: float, sigma: float):
        """
        Paramètres
        ----------
        S     : Prix du sous-jacent
        K     : Prix d'exercice (strike)
        T     : Temps à l'échéance (en années)
        r     : Taux sans risque (annualisé)
        sigma : Volatilité (annualisée)
        """
        self.S = S
        self.K = K
        self.T = T
        self.r = r
        self.sigma = sigma

        # Précalcul de d+ et d-
        self.d_plus = (
            (math.log(S / K) + (r + 0.5 * sigma**2) * T)
            / (sigma * math.sqrt(T))
        )
        self.d_minus = self.d_plus - sigma * math.sqrt(T)

    # === Prix ===
    def price(self) -> float:
        """C = S Φ(d₊) - K e^{-rT} Φ(d₋)"""
        return (
            self.S * norm.cdf(self.d_plus)
            - self.K * math.exp(-self.r * self.T) * norm.cdf(self.d_minus)
        )

    # === Greeks ===
    def delta(self) -> float:
        """Δ = Φ(d₊)"""
        return norm.cdf(self.d_plus)

    def gamma(self) -> float:
        """Γ = φ(d₊) / [σ S √T]"""
        return norm.pdf(self.d_plus) / (self.sigma * self.S * math.sqrt(self.T))

    def vega(self) -> float:
        """V = S √T φ(d₊)   (pour un shift de 1 en σ ; diviser par 100 pour 1%)"""
        return self.S * math.sqrt(self.T) * norm.pdf(self.d_plus)

    def theta(self) -> float:
        """Θ = -σS φ(d₊) / [2√T] - rK e^{-rT} Φ(d₋)   (par an)"""
        term1 = -self.sigma * self.S * norm.pdf(self.d_plus) / (2 * math.sqrt(self.T))
        term2 = -self.r * self.K * math.exp(-self.r * self.T) * norm.cdf(self.d_minus)
        return term1 + term2

    def rho(self) -> float:
        """ρ = K T e^{-rT} Φ(d₋)   (pour un shift de 1 en r ; diviser par 100 pour 1%)"""
        return (
            self.K * self.T * math.exp(-self.r * self.T) * norm.cdf(self.d_minus)
        )

    def summary(self) -> dict:
        """Retourne un dictionnaire avec tous les résultats."""
        return {
            "Prix du Call": self.price(),
            "Delta": self.delta(),
            "Gamma": self.gamma(),
            "Vega": self.vega(),
            "Theta (par an)": self.theta(),
            "Theta (par jour)": self.theta() / 365,
            "Rho": self.rho(),
        }


# ====================================================================
#                        EXEMPLE D'UTILISATION
# ====================================================================
if __name__ == "__main__":
    # Paramètres
    S = 100       # Spot
    K = 100       # Strike
    T = 1         # Maturité (1 an)
    r = 0.05      # Taux sans risque (5%)
    sigma = 0.2   # Volatilité (20%)

    bs = BlackScholes(S, K, T, r, sigma)

    print("=" * 55)
    print("  BLACK-SCHOLES : European Call Option Pricer")
    print("=" * 55)
    print(f"  S = {S}  |  K = {K}  |  T = {T}")
    print(f"  r = {r}  |  σ = {sigma}")
    print("-" * 55)
    print(f"  d₊            = {bs.d_plus:>12.6f}")
    print(f"  d₋            = {bs.d_minus:>12.6f}")
    print("-" * 55)

    results = bs.summary()
    for name, value in results.items():
        print(f"  {name:<18s} = {value:>12.6f}")

    print("=" * 55)
